server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java

```diff
@@ -383,27 +393,86 @@ protected Query doToQuery(QueryShardContext context) {
+            } else {
+                query = new ConstantScoreQuery(prefixTreeStrategy.makeQuery(getArgs(shapeToQuery, relation)));
+            }
         } else {
-            query = new ConstantScoreQuery(strategy.makeQuery(getArgs(shapeToQuery, relation)));
+            query = getVectorQuery(context, shapeFieldType, shapeToQuery);
         }
         return query;
     }
 
+    private Query getVectorQuery(QueryShardContext context, GeoShapeFieldMapper.GeoShapeFieldType shapeFieldType,
+                                 ShapeBuilder queryShapeBuilder) {
+        // CONTAINS queries are not yet supported by VECTOR strategy
+        if (relation == ShapeRelation.CONTAINS) {
+            throw new QueryShardException(context,
+                ShapeRelation.CONTAINS + " query relation not supported for Field [" + fieldName
+                    + "] when using Strategy [" + shapeFieldType.strategy() + "]");
+        }
+
+        // wrap geoQuery as a ConstantScoreQuery
+        return new ConstantScoreQuery(getVectorQuery(context, queryShapeBuilder.buildLucene()));
+    }
+
+    private Query getVectorQuery(QueryShardContext context, Object queryShape) {
+        Query geoQuery;
+        if (queryShape instanceof Line[]) {
+            geoQuery = LatLonShape.newLineQuery(fieldName(), relation.getLuceneRelation(), (Line[]) queryShape);
+        } else if (queryShape instanceof Polygon[]) {
+            geoQuery = LatLonShape.newPolygonQuery(fieldName(), relation.getLuceneRelation(), (Polygon[]) queryShape);
+        } else if (queryShape instanceof Line) {
+            geoQuery = LatLonShape.newLineQuery(fieldName(), relation.getLuceneRelation(), (Line) queryShape);
+        } else if (queryShape instanceof Polygon) {
+            geoQuery = LatLonShape.newPolygonQuery(fieldName(), relation.getLuceneRelation(), (Polygon) queryShape);
+        } else if (queryShape instanceof Rectangle) {
+            Rectangle r = (Rectangle) queryShape;
+            geoQuery = LatLonShape.newBoxQuery(fieldName(), relation.getLuceneRelation(),
+                r.minLat, r.maxLat, r.minLon, r.maxLon);
+        } else if (queryShape instanceof double[][]) {
+            // note: we decompose point queries into a bounding box query with min values == max values
+            // to do this for multipoint we would have to create a BooleanQuery for each point
+            // this is *way* too costly. So we do not allow multipoint queries
+            throw new QueryShardException(context, "Field [" + fieldName + "] does not support " + GeoShapeType.MULTIPOINT + " queries");
+        } else if (queryShape instanceof double[] || queryShape instanceof GeoPoint) {
+            // for now just create a single bounding box query with min values == max values
+            double[] pt = queryShape instanceof GeoPoint
+                ? new double[] {((GeoPoint)queryShape).lon(), ((GeoPoint)queryShape).lat()}
+                : (double[])queryShape;
```

Comment:
should we fail if the length of the `double[]` is not 2?

Reply:
:+1: